File: MPP/csharp/csharp/repository/TicketDatabaseRepositoryImpl.cs
using System.Data;
using csharp.domain;
using Npgsql;
namespace csharp.repository;

public class TicketDatabaseRepositoryImpl : AbstractDatabaseRepository<long, Ticket>, ITicketRepository
{

    private IConcertRepository _concertRepository;

    private IUserRepository _userRepository; 
    
    public TicketDatabaseRepositoryImpl(string url, IConcertRepository concertRepository, IUserRepository userRepository) : base(url)
    {
        _concertRepository = concertRepository;
        _userRepository = userRepository;
    }

    protected override Ticket ExtractEntity(NpgsqlDataReader resultSet)
    {
        return new Ticket(
            resultSet.GetInt64("id"),
            _concertRepository.FindOne(resultSet.GetInt64("concert_id")),
            resultSet.GetString("buyer"),
            resultSet.GetInt32("seats"),
            _userRepository.FindOne(resultSet.GetInt64("user_id"))
        );
    }

    protected override string GetFindOneQuery()
    {
        return "select * from tickets where id = @id";
    }

    protected override void SetIdParameter(NpgsqlCommand statement, long id)
    {
        statement.Parameters.AddWithValue("@id", id);
    }


    protected override string GetFindAllQuery()
    {
        return "select * from tickets";
    }


    protected override string GetSaveQuery()
    {
        return "insert into tickets (concert_id, buyer, seats, user_id) values (@concert_id, @buyer, @seats, @user_id)";
    }

    protected override void SetSaveQueryParameters(NpgsqlCommand statement, Ticket entity)
    {
        statement.Parameters.AddWithValue("@concert_id", entity.Concert.Id);
        statement.Parameters.AddWithValue("@buyer", entity.Buyer);
        statement.Parameters.AddWithValue("@seats", entity.Seats);
        statement.Parameters.AddWithValue("@user_id", entity.User.Id);
    }

    protected override string GetDeleteQuery()
    {
        return "delete from tickets where id = @id";
    }

    protected override string GetUpdateQuery()
    {
        return "update tickets set concert_id = @concert_id, buyer = @buyer, seats = @seats, user_id = @user_id where id = @id";
    }

    protected override void SetUpdateQueryParameters(NpgsqlCommand statement, Ticket entity)
    {
        statement.Parameters.AddWithValue("@id", entity.Id);
        statement.Parameters.AddWithValue("@concert_id", entity.Concert.Id);
        statement.Parameters.AddWithValue("@buyer", entity.Buyer);
        statement.Parameters.AddWithValue("@seats", entity.Seats);
        statement.Parameters.AddWithValue("@user_id", entity.User.Id);
    }
}

File: MPP/javafx/demo/src/main/java/org/example/repository/TicketDatabaseRepositoryImpl.java
package org.example.repository;

import org.example.domain.Ticket;

import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.util.Properties;

public class TicketDatabaseRepositoryImpl extends AbstractDatabaseRepository<Long, Ticket> implements TicketRepository {

    private final UserRepository userRepository;

    private final ConcertRepository concertRepository;

    public TicketDatabaseRepositoryImpl(UserRepository userRepository, ConcertRepository concertRepository, Properties properties) {
        super(properties);
        this.userRepository = userRepository;
        this.concertRepository = concertRepository;
    }

    @Override
    protected Ticket extractEntity(ResultSet resultSet) throws SQLException {
        return new Ticket(
                resultSet.getLong("id"),
                concertRepository.findOne(resultSet.getLong("concert_id")).orElseThrow(),
                userRepository.findOne(resultSet.getLong("user_id")).orElseThrow(),
                resultSet.getString("buyer"),
                resultSet.getInt("seats")
        );
    }

    @Override
    protected String getFindOneQuery() {
        return "select * from tickets where id = ?";
    }

    @Override
    protected void setIdParameter(PreparedStatement statement, Long aLong) throws SQLException {
        statement.setLong(1, aLong);
    }

    @Override
    protected String getFindAllQuery() {
        return "select * from tickets";
    }

    @Override
    protected String getSaveQuery() {
        return "insert into tickets (concert_id, user_id, buyer, seats) values (?, ?, ?, ?)";
    }

    @Override
    protected void setSaveParameters(PreparedStatement statement, Ticket entity) throws SQLException {
        statement.setLong(1, entity.getConcert().getId());
        statement.setLong(2, entity.getUser().getId());
        statement.setString(3, entity.getBuyer());
        statement.setInt(4, entity.getSeats());
    }

    @Override
    protected String getDeleteQuery() {
        return "delete from tickets where id = ?";
    }

    @Override
    protected String getUpdateQuery() {
        return "update tickets set concert_id = ?, user_id = ?, buyer = ?, seats = ? where id = ?";
    }

    @Override
    protected void setUpdateParameters(PreparedStatement statement, Ticket entity) throws SQLException {
        statement.setLong(1, entity.getConcert().getId());
        statement.setLong(2, entity.getUser().getId());
        statement.setString(3, entity.getBuyer());
        statement.setInt(4, entity.getSeats());
        statement.setLong(5, entity.getId());
    }
}
